File: Visual Studio Product Team/Official Visual Studio 2008 VB.NET Samples/46871-Official Visual Studio 2008 VB.NET Samples/Expression Tree Visualizer Sample (Visual Basic)/VB/ExpressionTreeVisualizer/TreeNode.vb
Imports System.Collections.ObjectModel
Imports System.Linq.Expressions
Imports System.Reflection
Imports System.Runtime.CompilerServices
Imports System.Runtime.Serialization
Imports System.Text
Imports System.Windows.Forms

<Extension()> _
Module ExpressionTreeExtension

    <Extension()> _
    Function ExtractName(ByVal name As String) As String
        Dim i = name.LastIndexOf("`")
        If i > 0 Then name = name.Substring(0, i)
        Return name
    End Function

    <Extension()> _
    Function ExtractGenericArguments(ByVal names() As Type) As String
        Dim builder = New StringBuilder("<")
        For Each genericArgument In names
            If builder.Length <> 1 Then builder.Append(", ")
            builder.Append(ObtainOriginalName(genericArgument))
        Next
        builder.Append(">")
        Return builder.ToString()
    End Function

    <Extension()> _
    Function ObtainOriginalName(ByVal type As Type) As String
        If Not type.IsGenericType Then
            Return type.Name
        Else
            Return ExtractName(type.Name) & ExtractGenericArguments(type.GetGenericArguments())
        End If
    End Function

    <Extension()> _
    Function ObtainOriginalMethodName(ByVal method As MethodInfo) As String
        If Not method.IsGenericMethod Then
            Return method.Name
        Else
            Return ExtractName(method.Name) & ExtractGenericArguments(method.GetGenericArguments())
        End If
    End Function

End Module

<Serializable()> _
Public Class ExpressionTreeNode
    Inherits TreeNode

    Public Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)
        MyBase.New(info, context)
    End Sub

    Public Sub New(ByVal value As Object)
        Dim type = value.GetType()
        Text = type.ObtainOriginalName()

        If TypeOf value Is Expression Then
            ImageIndex = 2
            SelectedImageIndex = 2

            Dim propertyInfos() As PropertyInfo = Nothing
            If type.IsGenericType AndAlso type.GetGenericTypeDefinition() Is GetType(Expression(Of )) Then
                propertyInfos = type.BaseType.GetProperties(BindingFlags.Public Or BindingFlags.Instance)
            Else
                propertyInfos = type.GetProperties(BindingFlags.Public Or BindingFlags.Instance)
            End If

            For Each propertyInfo In propertyInfos
                If propertyInfo.Name <> "nodeType" Then
                    Nodes.Add(New AttributeNode(value, propertyInfo))
                End If
            Next
        Else
            ImageIndex = 4
            SelectedImageIndex = 4
            Text = """" & value.ToString() & """"
        End If
    End Sub

    Protected Overrides Sub Serialize(ByVal si As SerializationInfo, ByVal context As StreamingContext)
        MyBase.Serialize(si, context)
    End Sub

End Class

<Serializable()> _
Public Class AttributeNode
    Inherits TreeNode

    Public Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)
        MyBase.new(info, context)
    End Sub

    Public Sub New(ByVal attribute As Object, ByVal propertyInfo As PropertyInfo)
        Text = propertyInfo.Name & " : " & propertyInfo.PropertyType.ObtainOriginalName()
        ImageIndex = 3
        SelectedImageIndex = 3

        Dim value = propertyInfo.GetValue(attribute, Nothing)
        If value IsNot Nothing Then
            If value.GetType().IsGenericType AndAlso value.GetType().GetGenericTypeDefinition() Is GetType(ReadOnlyCollection(Of )) Then
                If value.GetType().InvokeMember("get_Count", BindingFlags.InvokeMethod, Nothing, value, Nothing) = 0 Then
                    Text &= " : Empty"
                Else
                    For Each tree In value
                        If TypeOf tree Is Expression Then
                            Nodes.Add(New ExpressionTreeNode(tree))
                        ElseIf TypeOf tree Is MemberAssignment Then
                            Nodes.Add(New ExpressionTreeNode(CType(tree, MemberAssignment).Expression))
                        End If
                    Next
                End If
            ElseIf TypeOf value Is Expression Then
                Text &= CType(value, Expression).NodeType
                Nodes.Add(New ExpressionTreeNode(value))
            ElseIf TypeOf value Is MethodInfo Then
                Text &= " : """ & CType(value, MethodInfo).ObtainOriginalMethodName() & """"
            ElseIf TypeOf value Is Type Then
                Text &= " : """ & CType(value, Type).ObtainOriginalName() & """"
            Else
                Text &= " : """ & value.ToString() & """"
            End If
        Else
            Text &= " : Nothing"
        End If
    End Sub

    Protected Overrides Sub Serialize(ByVal si As SerializationInfo, ByVal context As StreamingContext)
        MyBase.Serialize(si, context)
    End Sub

End Class

File: Visual Studio Product Team/Official Visual Studio 2008 C# Samples/46335-Official Visual Studio 2008 C# Samples/ExpressionTreeVisualizer/C#/ExpressionTreeVisualizer/TreeNode.cs
using System;
using System.Text;
using System.Drawing;
using System.Globalization;
using System.IO;
using System.Reflection;
using System.Windows.Forms;
using Microsoft.CSharp;
using System.Collections;
using System.Linq.Expressions;
using Microsoft.VisualStudio.DebuggerVisualizers;
using System.Diagnostics;
using System.Runtime.Serialization;
using System.Collections.ObjectModel;
using System.Linq;
[assembly: CLSCompliant(true)]

namespace ExpressionVisualizer {
    static class ExpressionTreeExtention {
        static string ExtractName(string name)
        {
            int i = name.LastIndexOf("`", StringComparison.Ordinal);
            if (i > 0)
                name = name.Substring(0, i);
            return name;
        }

        static string ExtractGenericArguments(Type[] names)
        {
            StringBuilder builder = new StringBuilder("<");
            foreach (Type genericArgument in names) {
                if (builder.Length != 1) builder.Append(", ");
                builder.Append(ObtainOriginalName(genericArgument));
            }
            builder.Append(">");
            return builder.ToString();
        }

        public static string ObtainOriginalName(this Type type)
        {
            if (!type.IsGenericType) {
                return type.Name;
            }
            else {
                return ExtractName(type.Name) + ExtractGenericArguments(type.GetGenericArguments());
            }
        }

        public static string ObtainOriginalMethodName(this MethodInfo method)
        {
            if (!method.IsGenericMethod) {
                return method.Name;
            }
            else {
                return ExtractName(method.Name) + ExtractGenericArguments(method.GetGenericArguments());
            }
        }
    }

    [Serializable]
    public class ExpressionTreeNode : TreeNode {
        protected ExpressionTreeNode(SerializationInfo info, StreamingContext context) : base(info, context) { }

        protected override void Serialize(System.Runtime.Serialization.SerializationInfo si, System.Runtime.Serialization.StreamingContext context)
        {
            base.Serialize(si, context);
        }

        public ExpressionTreeNode(Object value)
        {
            Type type = value.GetType();
            Text = type.ObtainOriginalName();

            if (value is Expression) {
                ImageIndex = 2;
                SelectedImageIndex = 2;

                PropertyInfo[] propertyInfos = null;
                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Expression<>)) {
                    propertyInfos = type.BaseType.GetProperties(BindingFlags.Public | BindingFlags.Instance);
                }
                else {
                    propertyInfos = type.GetProperties(BindingFlags.Public | BindingFlags.Instance);
                }

                foreach (PropertyInfo propertyInfo in propertyInfos) {
                    if ((propertyInfo.Name != "nodeType")) {
                        Nodes.Add(new AttributeNode(value, propertyInfo));
                    }
                }
            }
            else {
                ImageIndex = 4;
                SelectedImageIndex = 4;
                Text = "\"" + value.ToString() + "\"";
            }
        }
    }

    [Serializable]
    public class AttributeNode : TreeNode {
        protected AttributeNode(SerializationInfo info, StreamingContext context) : base(info, context) { }
        public AttributeNode(Object attribute, PropertyInfo propertyInfo)
        {
            Text = propertyInfo.Name + " : " + propertyInfo.PropertyType.ObtainOriginalName();
            ImageIndex = 3;
            SelectedImageIndex = 3;

            Object value = propertyInfo.GetValue(attribute, null);
            if (value != null) {
                if (value.GetType().IsGenericType && value.GetType().GetGenericTypeDefinition() == typeof(ReadOnlyCollection<>)) {
                    if ((int)value.GetType().InvokeMember("get_Count", BindingFlags.InvokeMethod, null, value, null, CultureInfo.InvariantCulture) == 0) {
                        Text += " : Empty";
                    }
                    else {
                        foreach (object tree in (IEnumerable)value) {
                            if (tree is Expression) {
                                Nodes.Add(new ExpressionTreeNode(tree));
                            }
                            else if (tree is MemberAssignment) {
                                Nodes.Add(new ExpressionTreeNode(((MemberAssignment)tree).Expression));
                            }
                        }
                    }
                }
                else if (value is Expression) {
                    Text += ((Expression)value).NodeType;
                    Nodes.Add(new ExpressionTreeNode(value));
                }
                else if (value is MethodInfo) {
                    MethodInfo minfo = value as MethodInfo;
                    Text += " : \"" + minfo.ObtainOriginalMethodName() + "\"";
                }
                else if (value is Type) {
                    Type minfo = value as Type;
                    Text += " : \"" + minfo.ObtainOriginalName() + "\"";
                }
                else {
                    Text += " : \"" + value.ToString() + "\"";
                }
            }
            else {
                Text += " : null";
            }
        }
        protected override void Serialize(System.Runtime.Serialization.SerializationInfo si, System.Runtime.Serialization.StreamingContext context)
        {
            base.Serialize(si, context);
        }

    }
}